### PR TITLE
Optionally view workshop map with Mapbox

### PR DESCRIPTION
**Pull request by bethanyaconnor**
Using the `?mapbox` param in the URL, we can see the Mapbox map instead of the google map.

Screenshot 2020-10-02 at 12 52 41 PM

With popup:
Screenshot 2020-10-04 at 12 34 05 PM

mapbox-geocode

## Links

- [spec]()
- [jira]()

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/sites/code.org/pages/views/workshop_search.js [hunk 1]
```diff
@@ -10,30 +11,265 @@ const markerCustererOptions = {
...
+  var workshops = JSON.parse(e.features[0].properties['workshops']);
+  const description = compileMapboxPopupHtml(workshops, false);
+
+  new mapboxgl.Popup()
+    .setLngLat(coordinates)
+    .setHTML(description)
+    .addTo(map);
+}
+
+function compileMapboxPopupHtml(workshops, first) {
```
Comment (bethanyaconnor) [inline apps/src/sites/code.org/pages/views/workshop_search.js:214]: This is mostly copied from the Google Maps version but the data is structured slightly differently. Ideally we would deduplicate this code but we should be removing the google maps code soon.

File: apps/src/sites/code.org/pages/views/workshop_search.js [hunk 2]
```diff
@@ -10,30 +11,265 @@ const markerCustererOptions = {
...
+  map = new mapboxgl.Map({
+    container: 'mapbox-map',
+    style: 'mapbox://styles/mapbox/streets-v11',
+    center: [-95.665, 37.6],
+    zoom: 3
+  });
+
+  map.dragRotate.disable();
+  map.scrollZoom.disable();
+  map.dragPan.disable();
```
Comment (breville) [inline apps/src/sites/code.org/pages/views/workshop_search.js:44]: I guess the density & distribution of workshops make it okay to disable movement & zooming?

File: apps/src/sites/code.org/pages/views/workshop_search.js [hunk 3]
```diff
@@ -10,30 +11,265 @@ const markerCustererOptions = {
...
+    'bottom-right'
+  );
+  const enableMouseControls = () => {
+    map.scrollZoom.enable();
+    map.dragPan.enable();
+  };
+  // Enable mouse controls when the map is clicked
+  map.on('click', enableMouseControls);
+  // Enable mouse controls when the zoom (+/-) buttons are pressed
+  map.on('zoom', enableMouseControls);
```
Comment (breville) [inline apps/src/sites/code.org/pages/views/workshop_search.js:56]: Oh, I see.

File: apps/src/sites/code.org/pages/views/workshop_search.js [hunk 4]
```diff
@@ -10,30 +11,265 @@ const markerCustererOptions = {
...
+      });
+  });
+  map.on('mouseenter', 'workshop-clusters', function() {
+    map.getCanvas().style.cursor = 'pointer';
+  });
+}
+
+function placeIntroWorkshops() {
+  map.loadImage(
+    'https://maps.google.com/mapfiles/kml/paddle/red-blank.png',
```
Comment (breville) [inline apps/src/sites/code.org/pages/views/workshop_search.js:139]: To state the obvious, we should update this :)

File: apps/src/sites/code.org/pages/views/workshop_search.js [hunk 5]
```diff
@@ -10,30 +11,265 @@ const markerCustererOptions = {
...
+  map.on('click', 'intro-workshops', onMarkerClick);
+  map.on('mouseenter', 'intro-workshops', function() {
+    map.getCanvas().style.cursor = 'pointer';
+  });
+}
+
+function placeDeepDiveWorkshops() {
+  map.loadImage(
+    'https://maps.google.com/mapfiles/kml/paddle/red-stars.png',
+    (error, image) => {
```
Comment (breville) [inline apps/src/sites/code.org/pages/views/workshop_search.js:173]: In theory we could consolidate this into one retrieval.

File: dashboard/app/controllers/api/v1/pd/workshops_controller.rb [hunk 6]
```diff
@@ -108,6 +108,38 @@ def upcoming_teachercons
     render json: {error: e.message}, status: :bad_request
   end
 
+  def to_geojson(workshops)
```
Comment (breville) [inline dashboard/app/controllers/api/v1/pd/workshops_controller.rb:111]: Out of curiosity, why is this new function needed?

**Review by breville — APPROVED** [review]
Seems like a good update based on (very recent) Mapbox experience.

File: apps/src/sites/code.org/pages/views/workshop_search.js (hunk 5 above)
Comment (bethanyaconnor) [reply apps/src/sites/code.org/pages/views/workshop_search.js:173]: I tried and Mapbox was extremely uncooperative, so I split them 🤷

File: apps/src/sites/code.org/pages/views/workshop_search.js (hunk 4 above)
Comment (bethanyaconnor) [reply apps/src/sites/code.org/pages/views/workshop_search.js:139]: Agreed! I don't think this blocks this PR but let me know if you disagree!

File: apps/src/sites/code.org/pages/views/workshop_search.js (hunk 3 above)
Comment (bethanyaconnor) [reply apps/src/sites/code.org/pages/views/workshop_search.js:56]: Yeah, this was a thing on the hour of code map where a map that's just in the middle of a page with a bunch of other stuff will capture the scrolling. This forces the user to interact with the map to be able to scroll-zoom.

File: dashboard/app/controllers/api/v1/pd/workshops_controller.rb (hunk 6 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/api/v1/pd/workshops_controller.rb:111]: Mapbox works well with geojson (a json standard for locations) so I just pushed all the work off to that library when creating the sources/layers.

**Pull request merged**